File: Example/FWFramework/Classes/Module/Test/Plugin/TestAlertController.m
//
//  TestAlertController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestAlertController.h"
@import FWFramework;

@interface TestAlertController () <FWTableViewController>

@end

@implementation TestAlertController

- (UITableViewStyle)setupTableStyle
{
    return UITableViewStyleGrouped;
}

- (void)setupNavbar
{
    [self fw_setRightBarItem:@"切换插件" block:^(id  _Nonnull sender) {
        id<FWAlertPlugin> alertPlugin = [FWPluginManager loadPlugin:@protocol(FWAlertPlugin)];
        if (alertPlugin) {
            [FWPluginManager unloadPlugin:@protocol(FWAlertPlugin)];
            [FWPluginManager unregisterPlugin:@protocol(FWAlertPlugin)];
        } else {
            [FWPluginManager registerPlugin:@protocol(FWAlertPlugin) withObject:[FWAlertControllerImpl class]];
        }
    }];
}

- (void)setupSubviews
{
    [self.tableData addObjectsFromArray:@[
        @[@"警告框(简单)", @"onAlert1"],
        @[@"警告框(详细)", @"onAlert2"],
        @[@"确认框(简单)", @"onConfirm1"],
        @[@"确认框(详细)", @"onConfirm2"],
        @[@"输入框(简单)", @"onPrompt1"],
        @[@"输入框(详细)", @"onPrompt2"],
        @[@"输入框(复杂)", @"onPrompt3"],
        @[@"警告框(容错)", @"onAlertE"],
        @[@"操作表(简单)", @"onSheet1"],
        @[@"操作表(详细)", @"onSheet2"],
        @[@"弹出框(完整)", @"onAlertF"],
        @[@"弹出框(头部视图)", @"onAlertH"],
        @[@"弹出框(整个视图)", @"onAlertV"],
        @[@"警告框(样式)", @"onAlertA"],
        @[@"操作表(样式)", @"onSheetA"],
        @[@"警告框(优先)", @"onAlertP"],
        @[@"操作表(优先)", @"onSheetP"],
        @[@"关闭弹出框", @"onCloseA"],
    ]];
}

#pragma mark - TableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.tableData.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [UITableViewCell fw_cellWithTableView:tableView];
    NSArray *rowData = [self.tableData objectAtIndex:indexPath.row];
    cell.textLabel.text = [rowData objectAtIndex:0];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    NSArray *rowData = [self.tableData objectAtIndex:indexPath.row];
    SEL selector = NSSelectorFromString([rowData objectAtIndex:1]);
    if ([self respondsToSelector:selector]) {
        FWIgnoredBegin();
        [self performSelector:selector];
        FWIgnoredEnd();
    }
}

#pragma mark - Action

- (void)onAlert1
{
    [self fw_showAlertWithTitle:@"警告框标题"
                       message:@"警告框消息"
                        cancel:nil
                   cancelBlock:^{
                        NSLog(@"顶部控制器：%@", FWNavigator.topPresentedController);
                    }];
}

- (void)onAlert2
{
    [self fw_showAlertWithTitle:@"警告框标题"
                       message:@"警告框消息"
                         style:FWAlertStyleDefault
                        cancel:nil
                       actions:@[@"按钮1", @"按钮2"]
                   actionBlock:^(NSInteger index) {
                       NSLog(@"点击的按钮index: %@", @(index));
                   }
                   cancelBlock:^{
                       NSLog(@"点击了取消按钮");
                   }];
}

- (void)onConfirm1
{
    [self fw_showConfirmWithTitle:@"确认框标题"
                         message:@"确认框消息"
                          cancel:nil
                         confirm:nil
                    confirmBlock:^{
                        NSLog(@"点击了确定按钮");
                    }];
}

- (void)onConfirm2
{
    [self fw_showConfirmWithTitle:@"确认框标题"
                         message:@"确认框消息"
                          cancel:nil
                         confirm:@"我是很长的确定按钮"
                    confirmBlock:^{
                        NSLog(@"点击了确定按钮");
                    }
                     cancelBlock:^{
                         NSLog(@"点击了取消按钮");
                     }];
}

- (void)onPrompt1
{
    [self fw_showPromptWithTitle:@"输入框标题"
                        message:@"输入框消息"
                         cancel:nil
                        confirm:nil
                   confirmBlock:^(NSString *text){
                       NSLog(@"输入内容：%@", text);
                   }];
}

- (void)onPrompt2
{
    [self fw_showPromptWithTitle:@"输入框标题"
                        message:@"输入框消息"
                         cancel:nil
                        confirm:nil
                    promptBlock:^(UITextField *textField){
                        textField.placeholder = @"请输入密码";
                        textField.secureTextEntry = YES;
                    }
                   confirmBlock:^(NSString *text){
                       NSLog(@"输入内容：%@", text);
                   }
                    cancelBlock:^{
                        NSLog(@"点击了取消按钮");
                    }];
}

- (void)onPrompt3
{
    [self fw_showPromptWithTitle:@"输入框标题"
                        message:@"输入框消息"
                         cancel:nil
                        confirm:nil
                    promptCount:2
                    promptBlock:^(UITextField *textField, NSInteger index) {
                        if (index == 0) {
                            textField.placeholder = @"请输入用户名";
                            textField.secureTextEntry = NO;
                        } else {
                            textField.placeholder = @"请输入密码";
                            textField.secureTextEntry = YES;
                        }
                    }
                   confirmBlock:^(NSArray<NSString *> *values) {
                        NSLog(@"输入内容：%@", values);
                    }
                    cancelBlock:^{
                        NSLog(@"点击了取消按钮");
                    }];
}

- (void)onAlertE
{
    [self fw_showAlertWithTitle:nil
                       message:nil
                        cancel:nil
                   cancelBlock:^{
                        NSLog(@"顶部控制器：%@", FWNavigator.topPresentedController);
                    }];
}

- (void)onSheet1
{
    [self fw_showSheetWithTitle:nil
                       message:nil
                        cancel:@"取消"
                       actions:@[@"操作1"]
                   actionBlock:^(NSInteger index) {
                       NSLog(@"点击的操作index: %@", @(index));
                   }];
}

- (void)onSheet2
{
    [self fw_showSheetWithTitle:@"操作表标题"
                       message:@"操作表消息"
                        cancel:@"取消"
                       actions:@[@"操作1", @"操作2", @"操作3"]
                   currentIndex:1
                   actionBlock:^(NSInteger index) {
                       NSLog(@"点击的操作index: %@", @(index));
                   }
                   cancelBlock:^{
                       NSLog(@"点击了取消操作");
                   }];
}

- (void)onAlertF
{
    FWWeakifySelf();
    [self fw_showAlertWithTitle:@"请输入账号信息，我是很长很长很长很长很长很长的标题"
                       message:@"账户信息必填，我是很长很长很长很长很长很长的消息"
                         style:FWAlertStyleDefault
                        cancel:@"取消"
                       actions:@[@"重试", @"高亮", @"禁用", @"确定"]
                   promptCount:2
                   promptBlock:^(UITextField *textField, NSInteger index) {
                        if (index == 0) {
                            textField.placeholder = @"请输入用户名";
                            textField.secureTextEntry = NO;
                        } else {
                            textField.placeholder = @"请输入密码";
                            textField.secureTextEntry = YES;
                        }
                    }
                   actionBlock:^(NSArray<NSString *> *values, NSInteger index) {
                        FWStrongifySelf();
                        if (index == 0) {
                            [self onAlertF];
                        } else {
                            NSLog(@"输入内容：%@", values);
                        }
                    }
                   cancelBlock:^{
                        NSLog(@"点击了取消按钮");
                    }
                   customBlock:^(UIAlertController *alertController) {
                        alertController.preferredAction = alertController.actions[1];
                        alertController.actions[2].enabled = NO;
                        if ([alertController isKindOfClass:[FWAlertController class]]) {
                            ((FWAlertController *)alertController).image = [UIImage fw_appIconImage];
                        }
                    }];
}

- (void)onAlertH
{
    UIView *headerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 300, 200)];
    headerView.backgroundColor = UIColor.whiteColor;
    
    [[FWAlertControllerImpl sharedInstance] viewController:self
                                          showAlertWithStyle:UIAlertControllerStyleAlert
                                                  headerView:headerView
                                                 cancel:@"取消"
                                                actions:@[@"确定"]
                                            actionBlock:^(NSInteger index) {
                                                NSLog(@"点击了确定按钮");
                                            }
                                            cancelBlock:^{
                                                NSLog(@"点击了取消按钮");
                                            }
                                            customBlock:nil];
}

- (void)onAlertV
{
    UIView *alertView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 300, 200)];
    alertView.backgroundColor = UIColor.whiteColor;
    FWWeakifySelf();
    [alertView fw_addTapGestureWithBlock:^(id  _Nonnull sender) {
        FWStrongifySelf();
        [self.presentedViewController dismissViewControllerAnimated:YES completion:nil];
    }];
    
    [[FWAlertControllerImpl sharedInstance] viewController:self
                                          showAlertWithStyle:UIAlertControllerStyleAlert
                                                  headerView:alertView
                                                 cancel:nil
                                                actions:nil
                                            actionBlock:nil
                                            cancelBlock:nil
                                            customBlock:^(FWAlertController *alertController) {
                                                alertController.tapBackgroundViewDismiss = YES;
                                            }];
}

- (void)onAlertA
{
    NSMutableAttributedString *title = [NSMutableAttributedString new];
    NSTextAttachment *attachment = [NSTextAttachment new];
    attachment.image = [[UIImage fw_appIconImage] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    attachment.bounds = CGRectMake(0, -20, 30, 30);
    [title appendAttributedString:[NSAttributedString attributedStringWithAttachment:attachment]];
    NSDictionary *attrs = @{
        NSFontAttributeName: [UIFont fw_boldFontOfSize:17],
        NSForegroundColorAttributeName: [UIColor redColor],
    };
    [title appendAttributedString:[[NSAttributedString alloc] initWithString:@"\n\n警告框标题" attributes:attrs]];
        
    NSMutableAttributedString *message = [NSMutableAttributedString new];
    attrs = @{
        NSFontAttributeName: [UIFont fw_fontOfSize:15],
        NSForegroundColorAttributeName: [UIColor greenColor],
    };
    [message appendAttributedString:[[NSAttributedString alloc] initWithString:@"警告框消息" attributes:attrs]];
    
    [self fw_showAlertWithTitle:title
                       message:message
                         style:FWAlertStyleDefault
                        cancel:nil
                       actions:@[@"按钮1", @"按钮2", @"按钮3", @"按钮4"]
                   actionBlock:nil
                   cancelBlock:nil];
}

- (void)onSheetA
{
    NSMutableAttributedString *title = [NSMutableAttributedString new];
    NSTextAttachment *attachment = [NSTextAttachment new];
    attachment.image = [[UIImage fw_appIconImage] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    attachment.bounds = CGRectMake(0, -20, 30, 30);
    [title appendAttributedString:[NSAttributedString attributedStringWithAttachment:attachment]];
    NSDictionary *attrs = @{
        NSFontAttributeName: [UIFont fw_boldFontOfSize:17],
        NSForegroundColorAttributeName: [UIColor redColor],
    };
    [title appendAttributedString:[[NSAttributedString alloc] initWithString:@"\n\n操作表标题" attributes:attrs]];
        
    NSMutableAttributedString *message = [NSMutableAttributedString new];
    attrs = @{
        NSFontAttributeName: [UIFont fw_fontOfSize:15],
        NSForegroundColorAttributeName: [UIColor greenColor],
    };
    [message appendAttributedString:[[NSAttributedString alloc] initWithString:@"操作表消息" attributes:attrs]];
    
    [self fw_showSheetWithTitle:title
                       message:message
                        cancel:@"取消"
                       actions:@[@"操作1", @"操作2", @"操作3", @"操作4", @"操作5", @"操作6", @"操作7", @"操作8", @"操作9", @"操作10"]
                   actionBlock:^(NSInteger index) {
                       NSLog(@"点击的操作index: %@", @(index));
                   }];
}

- (void)onAlertP
{
    FWWeakifySelf();
    [self fw_showAlertWithTitle:@"高优先级" message:@"警告框消息" cancel:nil cancelBlock:^{
        FWStrongifySelf();
        [self fw_showAlertWithTitle:@"普通优先级" message:@"警告框消息" cancel:nil cancelBlock:^{
            FWStrongifySelf();
            [self fw_showAlertWithTitle:@"低优先级" message:@"警告框消息" cancel:nil cancelBlock:nil];
        }];
    }];
}

- (void)onSheetP
{
    FWWeakifySelf();
    [self fw_showSheetWithTitle:@"高优先级" message:@"操作表消息" cancel:nil cancelBlock:^{
        FWStrongifySelf();
        [self fw_showSheetWithTitle:@"普通优先级" message:@"操作表消息" cancel:nil cancelBlock:^{
            FWStrongifySelf();
            [self fw_showSheetWithTitle:@"低优先级" message:@"操作表消息" cancel:nil cancelBlock:nil];
        }];
    }];
}

- (void)onCloseA
{
    [self fw_showAlertWithTitle:nil message:@"我将在两秒后自动关闭"];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [FWNavigator.topPresentedController fw_showAlertWithTitle:nil message:@"我将在一秒后自动关闭"];
    });
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self fw_hideAlert:YES completion:nil];
    });
}

@end

File: Example/FWFramework/Classes/Module/Test/Plugin/TestAlertController.swift
//
//  TestAlertController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/15.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestAlertController: UIViewController, TableViewControllerProtocol {
    
    typealias TableElement = [String]
    
    func setupTableStyle() -> UITableView.Style {
        .grouped
    }
    
    func setupNavbar() {
        app.setRightBarItem(UIBarButtonItem.SystemItem.action.rawValue) { [weak self] _ in
            self?.app.showSheet(title: nil, message: nil, actions: ["切换按钮样式"], actionBlock: { index in
                if AlertAppearance.appearance.preferredActionColor == nil {
                    AlertAppearance.appearance.preferredActionColor = UIColor.red
                    AlertAppearance.appearance.preferredActionBlock = { vc in
                        return vc.actions.first
                    }
                    AlertControllerAppearance.appearance.preferredActionColor = UIColor.red
                    AlertControllerAppearance.appearance.preferredActionBlock = { vc in
                        return vc.actions.first
                    }
                } else {
                    AlertAppearance.appearance.preferredActionColor = nil
                    AlertAppearance.appearance.preferredActionBlock = nil
                    AlertControllerAppearance.appearance.preferredActionColor = nil
                    AlertControllerAppearance.appearance.preferredActionBlock = nil
                }
            })
        }
    }
    
    func setupSubviews() {
        tableData.append(contentsOf: [
            ["警告框(简单)", "onAlert1"],
            ["警告框(详细)", "onAlert2"],
            ["确认框(简单)", "onConfirm1"],
            ["确认框(详细)", "onConfirm2"],
            ["输入框(简单)", "onPrompt1"],
            ["输入框(详细)", "onPrompt2"],
            ["输入框(复杂)", "onPrompt3"],
            ["警告框(容错)", "onAlertE"],
            ["操作表(简单)", "onSheet1"],
            ["操作表(详细)", "onSheet2"],
            ["弹出框(完整)", "onAlertF"],
            ["弹出框(头部视图)", "onAlertH"],
            ["弹出框(整个视图)", "onAlertV"],
            ["警告框(样式)", "onAlertA"],
            ["操作表(样式)", "onSheetA"],
            ["警告框(优先)", "onAlertP"],
            ["操作表(优先)", "onSheetP"],
            ["操作表(多个)", "onSheetM"],
            ["关闭弹出框", "onCloseA"],
        ])
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        tableData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell.app.cell(tableView: tableView)
        let rowData = tableData[indexPath.row]
        cell.textLabel?.text = rowData[0]
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let rowData = tableData[indexPath.row]
        app.invokeMethod(NSSelectorFromString(rowData[1]))
    }
    
}

@objc extension TestAlertController {
    
    func onAlert1() {
        app.showAlert(title: "警告框标题", message: "警告框消息", cancel: nil) {
            UIWindow.app.showMessage(text: "顶部控制器：\(String(describing: Navigator.topPresentedController))")
        }
    }
    
    func onAlert2() {
        app.showAlert(title: "警告框标题", message: "警告框消息", style: .default, cancel: nil, actions: ["按钮1", "按钮2"]) { index in
            UIWindow.app.showMessage(text: "点击的按钮index: \(index)")
        } cancelBlock: {
            UIWindow.app.showMessage(text: "点击了取消按钮")
        }
    }
    
    func onConfirm1() {
        app.showConfirm(title: "确认框标题", message: "确认框消息", cancel: nil, confirm: nil) {
            UIWindow.app.showMessage(text: "点击了确定按钮")
        }
    }
    
    func onConfirm2() {
        app.showConfirm(title: "确认框标题", message: "确认框消息", cancel: nil, confirm: "我是很长的确定按钮") {
            UIWindow.app.showMessage(text: "点击了确定按钮")
        } cancelBlock: {
            UIWindow.app.showMessage(text: "点击了取消按钮")
        }
    }
    
    func onPrompt1() {
        app.showPrompt(title: "输入框标题", message: "输入框消息", cancel: nil, confirm: nil) { text in
            UIWindow.app.showMessage(text: "输入内容：\(text)")
        }
    }
    
    func onPrompt2() {
        app.showPrompt(title: "输入框标题", message: "输入框消息", cancel: nil, confirm: nil, promptBlock: { textField in
            textField.placeholder = "请输入密码"
            textField.isSecureTextEntry = true
        }) { text in
            UIWindow.app.showMessage(text: "输入内容：\(text)")
        } cancelBlock: {
            UIWindow.app.showMessage(text: "点击了取消按钮")
        }
    }
    
    func onPrompt3() {
        app.showPrompt(title: "输入框标题", message: "输入框消息", cancel: nil, confirm: nil, promptCount: 2, promptBlock: { textField, index in
            if index == 0 {
                textField.placeholder = "请输入用户名"
                textField.isSecureTextEntry = false
            } else {
                textField.placeholder = "请输入密码"
                textField.isSecureTextEntry = true
            }
        }) { values in
            UIWindow.app.showMessage(text: "输入内容：\(values)")
        } cancelBlock: {
            UIWindow.app.showMessage(text: "点击了取消按钮")
        }
    }
    
    func onAlertE() {
        app.showAlert(title: nil, message: nil, cancel: nil) {
            UIWindow.app.showMessage(text: "顶部控制器：\(String(describing: Navigator.topPresentedController))")
        }
    }
    
    func onSheet1() {
        app.showSheet(title: nil, message: nil, cancel: "取消", actions: ["操作1"], currentIndex: -1) { index in
            UIWindow.app.showMessage(text: "点击的操作index: \(index)")
        }
    }
    
    func onSheet2() {
        app.showSheet(title: "操作表标题", message: "操作表消息", cancel: "取消", actions: ["操作1", "操作2", "操作3"], currentIndex: 1) { index in
            UIWindow.app.showMessage(text: "点击的操作index: \(index)")
        } cancelBlock: {
            UIWindow.app.showMessage(text: "点击了取消操作")
        }
    }
    
    func onAlertF() {
        app.showAlert(title: "请输入账号信息，我是很长很长很长很长很长很长的标题", message: "账户信息必填，我是很长很长很长很长很长很长的消息", style: .default, cancel: "取消", actions: ["重试", "高亮", "禁用", "确定"], promptCount: 2) { textField, index in
            if index == 0 {
                textField.placeholder = "请输入用户名"
                textField.isSecureTextEntry = false
            } else {
                textField.placeholder = "请输入密码"
                textField.isSecureTextEntry = true
            }
        } actionBlock: { [weak self] values, index in
            if index == 0 {
                self?.onAlertF()
            } else {
                UIWindow.app.showMessage(text: "输入内容：\(values)")
            }
        } cancelBlock: {
            UIWindow.app.showMessage(text: "点击了取消按钮")
        } customBlock: { param in
            if let alert = param as? UIAlertController {
                alert.preferredAction = alert.actions[1]
                alert.actions[2].isEnabled = false
            } else if let alert = param as? AlertController {
                alert.preferredAction = alert.actions[1]
                alert.actions[2].isEnabled = false
                alert.image = UIImage.app.appIconImage()
            }
        }
    }
    
    func onAlertH() {
        let headerView = UIView(frame: CGRect(x: 0, y: 0, width: 300, height: 200))
        headerView.backgroundColor = .white
        
        AlertControllerImpl.shared.showAlert(style: .alert, headerView: headerView, cancel: "取消", actions: ["确定"], actionBlock: { index in
            UIWindow.app.showMessage(text: "点击了确定按钮")
        }, cancelBlock: {
            UIWindow.app.showMessage(text: "点击了取消按钮")
        }, in: self)
    }
    
    func onAlertV() {
        let alertView = UIView(frame: CGRect(x: 0, y: 0, width: 300, height: 200))
        alertView.backgroundColor = .white
        alertView.app.addTapGesture { [weak self] _ in
            self?.presentedViewController?.dismiss(animated: true)
        }
        
        AlertControllerImpl.shared.showAlert(style: .alert, headerView: alertView, cancel: nil, actions: nil, actionBlock: nil, cancelBlock: nil, customBlock: { param in
            if let alert = param as? AlertController {
                alert.tapBackgroundViewDismiss = true
            }
        }, in: self)
    }
    
    func onAlertA() {
        let title = NSMutableAttributedString()
        let attachment = NSTextAttachment()
        attachment.image = UIImage.app.appIconImage()?.withRenderingMode(.alwaysOriginal)
        attachment.bounds = CGRect(x: 0, y: -20, width: 30, height: 30)
        title.append(NSAttributedString(attachment: attachment))
        var attrs: [NSAttributedString.Key: Any] = [
            .font: UIFont.app.boldFont(ofSize: 17),
            .foregroundColor: UIColor.red,
        ]
        title.append(NSAttributedString(string: "\n\n警告框标题", attributes: attrs))
        
        let message = NSMutableAttributedString()
        attrs = [
            .font: UIFont.app.font(ofSize: 15),
            .foregroundColor: UIColor.green,
        ]
        message.append(NSAttributedString(string: "警告框消息", attributes: attrs))
        
        app.showAlert(title: title, message: message, style: .default, cancel: nil, actions: ["按钮1", "按钮2", NSAttributedString(string: "按钮3", attributes: [.foregroundColor: UIColor.red]), NSAttributedString(string: "按钮4", attributes: [.foregroundColor: UIColor.green])], actionBlock: nil, cancelBlock: nil)
    }
    
    func onSheetA() {
        let title = NSMutableAttributedString()
        let attachment = NSTextAttachment()
        attachment.image = UIImage.app.appIconImage()?.withRenderingMode(.alwaysOriginal)
        attachment.bounds = CGRect(x: 0, y: -20, width: 30, height: 30)
        title.append(NSAttributedString(attachment: attachment))
        var attrs: [NSAttributedString.Key: Any] = [
            .font: UIFont.app.boldFont(ofSize: 17),
            .foregroundColor: UIColor.red,
        ]
        title.append(NSAttributedString(string: "\n\n操作表标题", attributes: attrs))
        
        let message = NSMutableAttributedString()
        attrs = [
            .font: UIFont.app.font(ofSize: 15),
            .foregroundColor: UIColor.green,
        ]
        message.append(NSAttributedString(string: "操作表消息", attributes: attrs))
        
        app.showSheet(title: title, message: message, cancel: "取消", actions: ["操作1", "操作2", NSAttributedString(string: "操作3", attributes: [.foregroundColor: UIColor.red]), NSAttributedString(string: "操作4", attributes: [.foregroundColor: UIColor.green]), "操作5", "操作6", NSAttributedString(string: "操作7", attributes: [.foregroundColor: UIColor.red]), NSAttributedString(string: "操作8", attributes: [.foregroundColor: UIColor.green]), "操作9", "操作10"], currentIndex: -1) { index in
            UIWindow.app.showMessage(text: "点击的操作index: \(index)")
        }
    }
    
    func onAlertP() {
        let taskManager = TaskManager(maxConcurrentTaskCount: 1)
        
        let task = TaskOperation(onMainThread: true, queuePriority: .low) { [weak self] task in
            self?.app.showAlert(title: "低优先级", message: "警告框消息", cancel: nil, cancelBlock: {
                task.finish()
            })
        }
        
        let task2 = TaskOperation(onMainThread: true, queuePriority: .normal) { [weak self] task in
            self?.app.showAlert(title: "普通优先级", message: "警告框消息", cancel: nil, cancelBlock: {
                task.finish()
            })
        }
        
        let task3 = TaskOperation(onMainThread: true, queuePriority: .high) { [weak self] task in
            self?.app.showAlert(title: "高优先级", message: "警告框消息", cancel: nil, cancelBlock: {
                task.finish()
            })
        }
        
        taskManager.addTasks([task, task2, task3])
    }
    
    func onSheetP() {
        let taskManager = TaskManager(maxConcurrentTaskCount: 1, isSuspended: true)
        
        let task = TaskOperation(onMainThread: true, queuePriority: .low) { [weak self] task in
            self?.app.showSheet(title: "低优先级", message: "操作表消息", cancel: nil, cancelBlock: {
                task.finish()
            })
        }
        
        let task2 = TaskOperation(onMainThread: true, queuePriority: .normal) { [weak self] task in
            self?.app.showSheet(title: "普通优先级", message: "操作表消息", cancel: nil, cancelBlock: {
                task.finish()
            })
        }
        
        let task3 = TaskOperation(onMainThread: true, queuePriority: .high) { [weak self] task in
            self?.app.showSheet(title: "高优先级", message: "操作表消息", cancel: nil) {
                task.finish()
            }
        }
        
        taskManager.addTask(task)
        taskManager.addTask(task2)
        taskManager.addTask(task3)
        taskManager.isSuspended = false
    }
    
    func onSheetM() {
        Navigator.topPresentedController?.app.showSheet(title: "第一个操作表", message: nil, cancel: "取消", actions: ["操作"], actionBlock: nil)
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            Navigator.topPresentedController?.app.showSheet(title: "第二个操作表", message: nil, cancel: "取消", actions: ["操作"], actionBlock: nil)
        }
    }
    
    func onCloseA() {
        app.showAlert(title: nil, message: "我将在两秒后自动关闭")
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            Navigator.topPresentedController?.app.showAlert(title: nil, message: "我将在一秒后自动关闭")
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            self?.app.hideAlert(animated: true)
        }
    }
    
}
